src/charm.py

```diff
@@ -57,44 +67,90 @@ def _on_install(self, _):
-        if self._stored.provider_ready:
-            self.unit.status = ActiveStatus()
+    def _on_config_changed(self, event):
+        self._configure(event)
 
     def _on_loki_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API."""
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Define an initial Pebble layer configuration
+        self._configure(event)
+
+    def _on_alertmanager_change(self, event):
+        self._configure(event)
+
+    def _configure(self, event):
+        """Configure Loki"""
+        restart = False
+
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for Pebble ready")
+            return False
+
+        current_layer = self._container.get_plan().services
+        new_layer = self._build_pebble_layer
+
+        if current_layer != new_layer:
+            restart = True
+
+        config = self._loki_config()
+
+        if yaml.safe_load(self._stored.config) != config:
+            self._stored.config = yaml.dump(config)
+            self._container.push(LOKI_CONFIG, self._stored.config)
+            logger.info("Pushed new configuration")
+            restart = True
+
+        if restart:
+            self._container.add_layer(self._name, new_layer, combine=True)
+            self._container.restart(self._name)
+            logger.info("Loki (re)started")
+
+        self.unit.status = ActiveStatus()
+
+    @property
+    def _loki_command(self):
+        """Construct command to launch Loki.
+
+        Returns:
+            a sting consisting of Loki command and associated
+            command line options.
+        """
+        return f"/usr/bin/loki -target=all -config.file={LOKI_CONFIG}"
+
+    @property
+    def _build_pebble_layer(self):
+        """Construct the pebble layer.
+
+        Returns:
+            a Pebble layer specification for the Loki workload container.
+        """
         pebble_layer = {
             "summary": "Loki layer",
             "description": "pebble config layer for Loki",
             "services": {
                 "loki": {
                     "override": "replace",
                     "summary": "loki",
-                    "command": "/usr/bin/loki -target=all -config.file=/etc/loki/local-config.yaml",
+                    "command": self._loki_command,
                     "startup": "enabled",
                 },
             },
         }
-        # Add intial Pebble config layer using the Pebble API
-        container.add_layer("loki", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
+
+        return pebble_layer
 
     ##############################################
     #             UTILITY METHODS                #
     ##############################################
     def _provide_loki(self):
```

Comment:
seems to overlap `_provide_loki` in `loki.py`

Reply:
I see it here:
https://github.com/canonical/loki-operator/blob/73a363604231f951aee35eb452d9c6089e7b8d85/lib/charms/loki_k8s/v0/loki.py#L47

Reply:
It's the docstring ;-)

Reply:
I was probably up for too long at the time of writing :D